Handle LN Address username conflicts on new wallet registration flow

erdemyerebasmaz opened the pull request:
Fixes #260
Fixes #263
Fixes #318
Fixes #319

This PR introduces significant refactoring of the `WebhookCubit`, replacing it with the newly created `LnAddressCubit`. The changes are aimed at improving the management of Lightning Address setup and username customization, as well as enhancing the user experience when handling username conflicts.

Depends on:
- `breez-lnurl`: [Set HTTP code 409 on username conflict #11](https://github.com/breez/breez-lnurl/pull/11)
  - to handle username conflicts

#### Changelist:
- Store LN Address Setup status on secure storage
- Renamed secure storage functions for consistency with the Cubit
- Utilize profile username only when registering a new wallet
- Implemented retry and randomization logic to handle username conflicts during registration.
#### UI-related:
- Created a new username formatter and text input formatter to handle username formatting & validation
- Added a loading indicator on the "Done" button in the Update LN Address Username bottom sheet while the `LnAddressCubit` is loading.
- Separated the status of LN Address registration from the LN Address username update status to allow more control over error handling
  - Username conflict errors are now shown on the Update LN Address Username bottom sheet
  - Update LN Address Username bottom sheet isn't dismissed on conflict errors
#### Misc: 
- Added a factory class for creating instances of LnAddressCubit to improve dependency management.

erdemyerebasmaz commented:
Made additional changes to include username length validation, fixes on username sanitization, and consistent naming across the app & lnurl service for maintainability.

Additionally, error handling in `WebhookService` is improved, and retry logic in `LnUrlPayService` is refactored for reusability and the `LnAddressCubit` is reorganized for better readability.

### Changelist:
- Validate username length for LN Address 104edb8
- `chore:` renaming to be consistent with LNURL service 395c566
- `chore:` rename webhook registration status on secure storage 69beaaa
- `fix:` convert sanitized username to lowercase & remove all whitespace 0a97d35
- add error handling on `WebhookService` 4d3eab3
- refactor `LnUrlPayService`, extract retry logic to a function 99799e6
- refactor `LnAddressCubit` to improve readability be54971

danielgranhao reviewed (commented):
Looks good! Just a couple of comments

erdemyerebasmaz commented on lib/cubit/ln_address/ln_address_cubit.dart:
```diff
@@ -0,0 +1,238 @@
+    String? username,
+  }) async {
+    final RegisterLnurlPayRequest request = RegisterLnurlPayRequest(
+      time: DateTime.now().millisecondsSinceEpoch ~/ 1000,
```
Correct!

Fixed and made some changes so that `_registerLnurlWebhook` takes `RegisterLnurlPayRequest` as an argument.

- [fix: use the same timestamp for generating the signature & registration request](https://github.com/breez/misty-breez/pull/344/commits/3c45101635e3bce6f6a6bac7abe710f6cc2e5a7a)

erdemyerebasmaz commented on packages/breez_preferences/lib/src/breez_preferences.dart:
```diff
@@ -99,18 +100,18 @@ class BreezPreferences {
-  Future<String?> getLnUrlPayKey() async {
+  Future<String?> getWebhookUrl() async {
     final SharedPreferences prefs = await SharedPreferences.getInstance();
     return prefs.getString(_kLnUrlPayKey);
```
> Would it make sense to migrate the preference so that the key name matches? i.e. "lnurlpay_key" -> "webhook_url_key"

It will refresh users LN Addresses but since we're not live yet, I'm okay with this migration without adding any backwards compatibility.

Done. 
- [refactor: Remove unused values on BreezPreferences](https://github.com/breez/misty-breez/pull/344/commits/b00a981efb63b37320d1c2a5a59d139a65c4b202) b00a981efb63b37320d1c2a5a59d139a65c4b202

I've also refactored it a bit so the unused keys are removed.
- Created a getter for `SharedPreferences` instance
  - Replaced get functions w/ getters
- Added comments to section keys
- Added logs to each fn for debugging purposes
- Renamed keys & their fn's for consistency

This change will also cause a regression for `report_preference_key` as it's key & value has changed, but this is negligible.

erdemyerebasmaz commented:
Addressed feedback & made additional changes to ease the review process & code maintainability.

### Feedback changes:
- [fix: use the same timestamp for generating the signature & registrati…](https://github.com/breez/misty-breez/pull/344/commits/2e1c848dfd22a8e3bb2d9964d60d6c51dacdfa3d) 2e1c848
- [refactor: Remove unused values on BreezPreferences](https://github.com/breez/misty-breez/pull/344/commits/718b17da8ba635ddd25742c20803c52dcf6cd65f)  718b17d
  - Created a getter for `SharedPreferences` instance
    - Used getters for get functions
  - Added comments to section keys
  - Renamed keys & their fn's for consistency
  - **This change will cause a negligible regression as the key value has changed for `_kBugReportBehavior`.**

### Changelist:
- [Add TODO comment for _registerWithRetries](https://github.com/breez/misty-breez/pull/344/commits/fb3acbf42b9095ed8138e999887cb14cd23a8311) fb3acbf
- [refactor: improve readability of setupLightningAddress function](https://github.com/breez/misty-breez/pull/344/commits/2b036eea4898cc15b5082e15adae986cce177228)  2b036ee
- [refactor: improve readability of _resolveUsername](https://github.com/breez/misty-breez/pull/344/commits/89df12ff4a0c8712fd2f1966c06ccd85bea30ff0) 89df12f
- [docs: Add docstrings for LnAddressCubit functions](https://github.com/breez/misty-breez/pull/344/commits/72701f4dabd21104fc70d8b82b423124810d7e17) 72701f4

### TODO: 
- [ ] Add docstrings for `LnUrlPayService` & `WebhookService`

dangeross reviewed (approved):
LGTM
